In [1]:
import numpy as np
import sympy
from sympy import simplify

titik_x = eval(input('Masukkan list titik-titik berurutan : ')) #nilai x
titik_y = eval(input('Masukkan list nilai fungsi di titik-titik tersebut : ')) #nilai y
x = eval(input('Masukkan titik dimana nilai turunan fungsinya akan diaproksimasi : ')) #titik x yang akan dicari turunannya
h = titik_x[1]-titik_x[0]

#Gunakan metode Interpolasi Lagrange
def persamaanpolinom(x_points, y_points):
    x=sympy.symbols('x')
    if  len(x_points)!= len(y_points):
        return 1
    pol=0
    for k in range(len(x_points)):
        L=1
        for j in range(len(x_points)):
            if j != k:
                L=L*((x-x_points[j]) /(x_points[k]-x_points[j]))
        pol+= L*y_points[k]
    return pol
polinom = persamaanpolinom(titik_x,titik_y)
Fx = str(simplify(polinom))

print(' ')
print('Fungsi Polinomial = :',Fx)
print(' ')

def f(x) : #mendefinisikan fungsi hasil polinom lagrange
    return eval(Fx)

print('1. Metode Three Point Endpoint\n2. Metode Three Point Midpoint\n3. Metode Five Point Endpoint\n4. Metode Five Point Midpoint')
print(' ')
metode = eval(input('Pilih metode berdasarkan angka pilihan di atas : ')) #pilih metode
n = eval(input('Turunan ke berapa yang diinginkan : ')) #pilih turunan

if metode == 1:
    def TPEP(x,h,n): #pendefinisan fungsi Three Point Endpoint dengan n-kali penurunan              
        if n == 1:  #turunan pertama
            return ((1/(2*h))*(-3*f(x)+4*f(x+h)-f(x+2*h)))
        elif n >= 2: #turunan ke-2 sampai ke-n
            return ((1/(2*h))*(-3*TPEP(x,h,n-1)+4*TPEP(x+h,h,n-1)-TPEP(x+2*h,h,n-1)))
    print('Turunan ke- {0} dari f di x0 = {1:.7f} adalah : '.format(n,x))
    print("f'({0}) = {1:.7f} (Three Point Endpoint)".format(x,TPEP(x,h,n)))

if metode == 2:
    def TPMP(x,h,n): #pendefinisan fungsi Three Point Midpoint dengan n-kali penurunan
        if n == 1:
            return (1/(2*h))*(f(x+h)-f(x-h))
        elif n >= 2:
            return (1/(2*h))*(TPMP(x+h,h,n-1)-TPMP(x-h,h,n-1))
    print('Turunan ke- {0} dari f di x0 = {1:.7f} adalah : '.format(n,x))
    print("f'({0}) = {1:.7f} (Three Point Midpoint)".format(x,TPMP(x,h,n)))

if metode == 3:
    def FPEP(x,h,n): #pendefinisan fungsi Five Point Endpoint dengan n-kali penurunan
        if n == 1:
            return (1/(12*h))*(-25*f(x)+48*f(x+h)-36*f(x+2*h)+16*f(x+3*h)-3*f(x+4*h))
        elif n >= 2:
            return (1/(12*h))*(-25*FPEP(x,h,n-1)+48*FPEP(x+h,h,n-1)-36*FPEP(x+2*h,h,n-1)+16*FPEP(x+3*h,h,n-1)-3*FPEP(x+4*h,h,n-1))
    print('Turunan ke- {0} dari f di x0 = {1:.7f} adalah : '.format(n,x))
    print("f'({0}) = {1:.7f} (Five Point Endpoint)".format(x,FPEP(x,h,n)))

if metode == 4:
    def FPMP(x,h,n): #pendefinisan fungsi Five Point Midpoint dengan n-kali penurunan
        if n == 1:
            return (1/(12*h))*(f(x-2*h)-8*f(x-h)+8*f(x+h)-f(x+2*h))
        elif n >= 2:
            return (1/(12*h))*(FPMP(x-2*h,h,n-1)-8*FPMP(x-h,h,n-1)+8*FPMP(x+h,h,n-1)-FPMP(x+2*h,h,n-1))
    print('Turunan ke- {0} dari f di x0 = {1:.7f} adalah : '.format(n,x))
    print("f'({0}) = {1:.7f} (Five Point Midpoint)".format(x,FPMP(x,h,n)))

Masukkan list titik-titik berurutan : [1.8,1.9,2.0,2.1,2.2]
Masukkan list nilai fungsi di titik-titik tersebut : [10.889365,12.703199,14.778112,17.148957,19.855030]
Masukkan titik dimana nilai turunan fungsinya akan diaproksimasi : 2
 
Fungsi Polinomial = : 1.85124999999607*x**4 - 8.63091666658875*x**3 + 22.1335874997312*x**2 - 22.0363508333685*x + 9.74379700009013
 
1. Metode Three Point Endpoint
2. Metode Three Point Midpoint
3. Metode Five Point Endpoint
4. Metode Five Point Midpoint
 
Pilih metode berdasarkan angka pilihan di atas : 1
Turunan ke berapa yang diinginkan : 1
Turunan ke- 1 dari f di x0 = 2.0000000 adalah : 
f'(2) = 22.0323100 (Three Point Endpoint)


In [ ]:
TPEP(2.7,0.1,3) #f'(2.4) untuk three point endpoint

-0.7849997393805097

In [ ]:
TPMP(2.4,0.1,1) #f'(2.4) untuk three point midpoint

1.145850123634773

In [ ]:
FPEP(2.4,0.1,1) #f'(2.4) untuk five point endpoint

1.1605001234314234

In [ ]:
FPMP(2.4,0.1,1) #f'(2.4) untuk five point midpoint

1.1572001234571871

In [ ]:
TPEP(2.4,0.1,2) #f''(2.5) untuk three point endpoint

-1.0499998714976755

In [ ]:
TPMP(2.4,0.1,2) #f''(2.5) untuk three point midpoint

-0.5169998706151091

In [ ]:
FPEP(2.4,0.1,2) #f''(2.5) untuk five point endpoint

-0.8721109777841062

In [ ]:
FPMP(2.4,0.1,2) #f''(2.5) untuk five point midpoint

-0.8054443153500617

In [ ]:
TPEP(2.7,0.1,3) #f'''(2.7) untuk three point endpoint

-0.7849997393805097

In [ ]:
TPMP(2.7,0.1,3) #f'''(2.7) untuk three point midpoint

0.7450001330653322

In [ ]:
FPEP(2.7,0.1,3) #f'''(2.7) untuk five point endpoint

0.7350012344799468

In [ ]:
FPMP(2.7,0.1,3) #f'''(2.7) untuk five point midpoint

0.7350001313073783

Program ini untuk mencari turunan numerik pertama, kedua, dan ketiga dari pasangan titik x dan y yang diketahui. Program ini menggunakan 4 metode turunan numerik yaitu Three Point Endpoint, Three Point Midpoint, Five Point Endpoint, Five Point Midpoint.
Algoritma program :
1. Menginput titik x dan titik y berurutan, nilai x yang akan dicari turunanya, dan stepsize.
2. Menggunakan interpolasi (pada program ini digunakan metode Lagrange) untuk mencari fungsi persamaan polinom.
3. Mendefinisikan fungsi f(x) untuk fungsi polinomial lagrange
4. Menginput metode yang dipilih dan turunan ke berapa yang diinginkan.
5. Gunakan if conditional untuk metode dan turunannya.
6. Jika metode dipilih 1 (Three Point Endpoint), definisikan fungsi TPEP dengan argumen x,h,n

   Masukan nilai fungsi TPEP $\frac{1}{2h}{[-3f(x)+4f(x+h)-f(x+2h)]}$ untuk turunan pertamanya
   
   dan nilai fungsi $\frac{1}{2h}[-3TPEP(x,h,n-1)+4TPEP(x+h,h,n-1)-TPEP(x+2h,h,n-1)]$ untuk turunan ke-2 dan ke-3.


7. Jika metode dipilih 2 (Three Point Midpoint), definisikan fungsi TPMP dengan argumen x,h,n

   Masukkan nilai fungsi TPMP $\frac{1}{2h}[f(x+h)-f(x-h)]$ untuk turunan pertamanya
   
   dan nilai fungsi $\frac{1}{2h}[TPMP(x+h,h,n-1)-TPMP(x-h,h,n-1)]$ untuk turunan ke-2 dan ke-3.
   
   
8. Jika metode dipilih 3 (Five Point Endpoint), definisikan fungsi FPEP dengan argumen x,h,n

   Masukkan nilai fungsi FPEP $\frac{1}{12h}[-25f(x)+48f(x+h)-36f(x+2h)+16f(x+3h)-3f(x+4h)]$ untuk turunan pertamanya
   
   dan nilai fungsi untuk turunan ke-2 dan ke-3
   $\frac{1}{12h}[-25FPEP(x,h,n-1)+48FPEP(x+h,h,n-1)-36FPEP(x+2h,h,n-1)+16FPEP(x+3h,h,n-1)-3FPEP(x+4h,h,n-1)]$
   
   
9. Jika metode dipilih 4 (Five Point Midpoint), definisikan fungsi FPEP dengan argumen x,h,n

   Masukkan nilai fungsi $\frac{1}{12h}[f(x-2h)-8f(x-h)+8f(x+h)-f(x+2h)]$ untuk turunan pertamanya
   
   dan nilai fungsi untuk turunan ke-2 dan ke-3
   $\frac{1}{12h}[FPMP(x-2h,h,n-1)-8FPMP(x-h,h,n-1)+8FPMP(x+h,h,n-1)-FPMP(x+2h,h,n-1)]$
   
   
 10. Print hasil turunan dari x yang dicari.

Jadi dari program yang telah dibuat dengan input nilai numerik sebagai berikut :
x = [2.4],[2.5],[2.6],[2.7],[2.8],[2.9],[3.0]
y = [1.76601],[1.87697],[1.97803],[2.06988],[2.15357],[2.22953],[2.29859]

Didapatkan nilai turunan pertama untuk x = 2.4 sebagai berikut :
1. Metode Three Point Endpoint f'(2.4) = 1.1591001230453912
2. Metode Three Point Midpoint f'(2.4) = 1.145850123634773
3. Metode Five Point Endpoint f'(2.4) = 1.1605001234314234
4. Metode Five Point Midpoint f'(2.4) = 1.1572001234571871

Didapatkan nilai turunan kedua untuk x = 2.5 sebagai berikut :
1. Metode Three Point Endpoint f''(2.5) = -1.0499998714976755
2. Metode Three Point Midpoint f''(2.5) = -0.5169998706151091
3. Metode Five Point Endpoint f''(2.5) = -0.8721109777841062
4. Metode Five Point Midpoint f''(2.5) = -0.8054443153500617

Didapatkan nilai turunan ketiga untuk x = 2.7 sebagai berikut :
1. Metode Three Point Endpoint f'''(2.7) = -0.7849997393805097
2. Metode Three Point Midpoint f'''(2.7) = 0.7450001330653322
3. Metode Five Point Endpoint f'''(2.7) = 0.7350012344799468
4. Metode Five Point Midpoint f'''(2.7) = 0.7350001313073783